src/metabase/task/sync_databases.clj

```diff
@@ -1,49 +1,169 @@
-            [clojure.tools.logging :as log]
+  "Scheduled tasks for syncing metadata/analyzing and caching FieldValues for connected Databases."
+  (:require [clojure.tools.logging :as log]
             [clojurewerkz.quartzite
+             [conversion :as qc]
              [jobs :as jobs]
              [triggers :as triggers]]
             [clojurewerkz.quartzite.schedule.cron :as cron]
             [metabase
-             [driver :as driver]
-             [sync :as sync]
-             [task :as task]]
+             [task :as task]
+             [util :as u]]
             [metabase.models.database :refer [Database]]
-            [toucan.db :as db]))
-
-(def ^:private ^:const sync-databases-job-key     "metabase.task.sync-databases.job")
-(def ^:private ^:const sync-databases-trigger-key "metabase.task.sync-databases.trigger")
-
-(defonce ^:private sync-databases-job (atom nil))
-(defonce ^:private sync-databases-trigger (atom nil))
-
-;; simple job which looks up all databases and runs a sync on them
-(jobs/defjob SyncDatabases [_]
-  (doseq [database (db/select Database, :is_sample false)] ; skip Sample Dataset DB
-    (try
-      ;; NOTE: this happens synchronously for now to avoid excessive load if there are lots of databases
-      ;; most of the time we do a quick sync and avoid the lengthy analysis process
-      ;; at midnight we run the full sync
-      (let [full-sync? (not (and (zero? (t/hour (t/now)))
-                                 (driver/driver-supports? (driver/engine->driver (:engine database)) :dynamic-schema)))]
-        (sync/sync-database! database {:full-sync? full-sync?}))
-      (catch Throwable e
-        (log/error (format "Error syncing database %d: " (:id database)) e)))))
+            [metabase.sync
+             [analyze :as analyze]
+             [field-values :as field-values]
+             [sync-metadata :as sync-metadata]]
+            [metabase.util.schema :as su]
+            [schema.core :as s]
+            [toucan.db :as db])
+  (:import metabase.models.database.DatabaseInstance
+           [org.quartz CronTrigger JobDetail JobKey TriggerKey]))
+
+;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; |                                                       JOB LOGIC                                                        |
+;;; +------------------------------------------------------------------------------------------------------------------------+
+
+(s/defn ^:private ^:always-validate job-context->database :- DatabaseInstance
+  "Get the Database referred to in JOB-CONTEXT. Guaranteed to return a valid Database."
+  [job-context]
+  (Database (u/get-id (get (qc/from-job-data job-context) "db-id"))))
+
+
+(jobs/defjob SyncAndAnalyzeDatabase [job-context]
+  (let [database (job-context->database job-context)]
+    (sync-metadata/sync-db-metadata! database)
+    ;; only run analysis if this is a "full sync" database
```

Comment:
If `is_full_sync` was off since the database was first added and then gets turned on then the current behavior is those Fields would get analyzed after the next sync process is ran, and would thereafter be considered "analyzed" and not be analyzed again until further notice. If we'd rather it happen as soon as someone changes the flag I can make that change as a follow-on

I haven't put too much thought into what should happen for manually triggering Field analysis since that's still several steps from here. But I don't see any reason why we couldn't still let people trigger manual analysis for a Field even if analysis is off by default for everything else (`is_full_sync` is `false`)